File: library/src/test/java/io/github/anderscheow/validator/conditions/ConditionTest.java
package io.github.anderscheow.validator.conditions;

import android.support.annotation.StringRes;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import io.github.anderscheow.validator.rules.common.MaxRule;
import io.github.anderscheow.validator.rules.common.MinRule;

import static org.junit.Assert.*;

public class ConditionTest {

    private Condition condition;

    @Before
    public void setUp() throws Exception {
    }

    @After
    public void tearDown() throws Exception {
        condition = null;
    }

    @Test
    public void add_OneBaseRule() throws Exception {
        condition = new Condition() {
            @Override
            public boolean validate(Object value) {
                return false;
            }
        }.add(new MinRule(5));

        assertEquals(1, condition.getBaseRules().size());
    }

    @Test
    public void add_TwoBaseRule() throws Exception {
        condition = new Condition() {
            @Override
            public boolean validate(Object value) {
                return false;
            }
        }.add(new MinRule(5))
            .add(new MaxRule(10));

        assertEquals(2, condition.getBaseRules().size());
    }

    @Test
    public void errorMessage_DefaultErrorMessage() throws Exception {
        String errorMessage = "Invalid input";

        condition = new Condition() {
            @Override
            public boolean validate(Object value) {
                return false;
            }
        };

        assertEquals(errorMessage, condition.errorMessage());
    }

    @Test
    public void errorMessage_CustomErrorMessage() throws Exception {
        String errorMessage = "This is a custom error message";

        condition = new Condition(errorMessage) {
            @Override
            public boolean validate(Object value) {
                return false;
            }
        };

        assertEquals(errorMessage, condition.errorMessage());
    }

    @Test
    public void errorMessage_CustomErrorRes() throws Exception {
        @StringRes int errorRes = 0;

        condition = new Condition(errorRes) {
            @Override
            public boolean validate(Object value) {
                return false;
            }
        };

        assertEquals(errorRes, condition.errorRes());
    }
}

File: library/src/test/java/io/github/anderscheow/validator/conditions/ConditionTest.kt
package io.github.anderscheow.validator.conditions

import androidx.annotation.StringRes
import io.github.anderscheow.validator.rules.common.MaxRule
import io.github.anderscheow.validator.rules.common.MinRule
import org.junit.After
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test

class ConditionTest {

    private lateinit var condition: Condition

    @Before
    @Throws(Exception::class)
    fun setUp() {
    }

    @After
    @Throws(Exception::class)
    fun tearDown() {
    }

    @Test
    @Throws(Exception::class)
    fun add_OneBaseRule() {
        condition = object : Condition(MinRule(5, ""), "") {
            override fun validate(value: String?): Boolean {
                return false
            }
        }

        assertEquals(1, condition.rules.size.toLong())
    }

    @Test
    @Throws(Exception::class)
    fun add_TwoBaseRule() {
        condition = object : Condition(listOf(MinRule(5, ""), MaxRule(10, "")), "") {
            override fun validate(value: String?): Boolean {
                return false
            }
        }

        assertEquals(2, condition.rules.size.toLong())
    }

    @Test
    @Throws(Exception::class)
    fun addAll_TwoBaseRule() {
        condition = object : Condition(listOf(MinRule(5, ""), MaxRule(10, "")), "") {
            override fun validate(value: String?): Boolean {
                return false
            }
        }

        assertEquals(2, condition.rules.size.toLong())
    }

    @Test
    @Throws(Exception::class)
    fun errorMessage_CustomErrorMessage() {
        val errorMessage = "This is a custom error message"

        condition = object : Condition(MinRule(5, ""), errorMessage) {
            override fun validate(value: String?): Boolean {
                return false
            }
        }

        assertEquals(errorMessage, condition.errorString)
        assertTrue(condition.isErrorAvailable)
        assertTrue(condition.isErrorMessageAvailable)
        assertFalse(condition.isErrorResAvailable)
    }

    @Test
    @Throws(Exception::class)
    fun errorMessage_CustomErrorRes() {
        @StringRes val errorRes = 0

        condition = object : Condition(MinRule(5, ""), errorRes) {
            override fun validate(value: String?): Boolean {
                return false
            }
        }

        assertEquals(errorRes, condition.errorRes)
        assertTrue(condition.isErrorAvailable)
        assertTrue(condition.isErrorMessageAvailable)
        assertTrue(condition.isErrorResAvailable)
    }
}